docs/CoreConfig.md
# Core Configuration Files

See [Logging](Logging.md).

The Haraka core reads some configuration files to determine a few actions:

- smtp.yaml or smtp.json

If either of these files exist then they are loaded first.
This file is designed to use the JSON/YAML file overrides documented in
[haraka-config](https://github.com/haraka/haraka-config) to optionally provide the entire configuration in a single file.

- databytes

Contains the maximum SIZE of an email that Haraka will receive.

- plugins

The list of plugins to load

- smtp.ini

  Keys:

  - listen_host, port - the host and port to listen on (default: ::0 and 25)
  - listen - (default: [::0]:25) Comma separated IP:Port addresses to listen on
  - inactivity_time - how long to let clients idle in seconds (default: 300)
  - nodes - specifies how many processes to fork. The string "cpus" will fork as many children as there are CPUs (default: 1, which enables cluster mode with a single process)
  - user - optionally a user to drop privileges to. Can be a string or UID.
  - group - optionally a group to drop privileges to. Can be a string or GID.
  - ignore_bad_plugins - If a plugin fails to compile by default Haraka will stop at load time.
    If, however, you wish to continue on without that plugin's facilities, then
    set this config option
  - daemonize - enable this to cause Haraka to fork into the background on start-up (default: 0)
  - daemon_log_file - (default: /var/log/haraka.log) where to redirect stdout/stderr when daemonized
  - daemon_pid_file - (default: /var/run/haraka.pid) where to write a PID file to
  - spool_dir - (default: none) directory to create temporary spool files in
  - spool_after - (default: -1) if message exceeds this size in bytes, then spool the message to disk
    specify -1 to disable spooling completely or 0 to force all messages to be spooled to disk.
  - graceful_shutdown - (default: false) enable this to wait for sockets on shutdown instead of closing them quickly
  - force_shutdown_timeout - (default: 30) number of seconds to wait for a graceful shutdown
  - smtputf8 - (default: true) advertise support for SMTPUTF8
  - strict_rfc1869 - (default: false) Requires senders to conform to RFC 1869 and RFC 821 when sending the MAIL FROM and RCPT TO commands. In particular,
    the inclusion of spurious spaces or missing angle brackets will be rejected.

- me

  A name to use for this server. Used in received lines and elsewhere. Setup
  by default to be your hostname.

- deny_includes_uuid

  Each connection and mail in Haraka includes a UUID which is also in most log
  messages. If you put a `1` in this file then every denied mail (either via
  DENY/5xx or DENYSOFT/4xx return codes) will include the uuid at the start
  of each line of the deny message in brackets, making it easy to track
  problems back to the logs.

  Because UUIDs are long, if you put a number greater than 1 in the config
  file, it will be truncated to that length. We recommend a 6 as a good
  balance of finding in the logs and not making lines too long.

- banner_includes_uuid

  This will add the full UUID to the first line of the SMTP greeting banner.

- early_talker_delay

  If clients talk early we _punish_ them with a delay of this many milliseconds
  default: 1000.

- plugin_timeout

  Seconds to allow a plugin to run before the next hook is called automatically
  default: 30

  Note also that each plugin can have a `config/<plugin_name>.timeout`
  file specifying a per-plugin timeout. In this file you can set a timeout of 0 to mean that this plugin's hooks never time out. Use this with care.

  If the plugin is in a sub-directory of plugins, then you must create this file
  in the equivalent path e.g. the queue/smtp_forward would need a timeout file in `config/queue/smtp_forward.timeout`

- smtpgreeting

  The greeting line used when a client connects. This can be multiple lines
  if required (this may cause some connecting machines to fail - though
  usually only spam-bots).

- max_received_count

  The maximum number of "Received" headers allowed in an email. This is a
  simple protection against mail loops. Defaults to 100.

- max_line_length

  The maximum length of lines in SMTP session commands (e.g. RCPT, HELO etc).
  Defaults to 512 (bytes) which is mandated by RFC 5321 §4.5.3.1.4. Clients
  exceeding this limit will be immediately disconnected with a "521 Command
  line too long" error.

- max_data_line_length

  The maximum length of lines in the DATA section of emails. Defaults to 992
  (bytes) which is the limit set by Sendmail. When this limit is exceeded the
  three bytes "\r\n " (0x0d 0x0a 0x20) are inserted into the stream to "fix"
  it. This has the potential to "break" some email, but makes it more likely
  to be accepted by upstream/downstream services, and is the same behaviour
  as Sendmail. Also when the data line length limit is exceeded
  `transaction.notes.data_line_length_exceeded` is set to `true`.

- outbound.concurrency_max

  Maximum concurrency to use when delivering mails outbound. Defaults to 100.

- outbound.disabled

  Put a `1` in this file to temporarily disable outbound delivery. Useful
  while figuring out network issues or testing.

- outbound.bounce_message

  The bounce message if delivery of the message fails. The default is normally fine. Bounce messages contain a number of template
  replacement values which are best discovered by looking at the source code.

- haproxy_hosts

  A list of HAProxy hosts that Haraka should enable the PROXY protocol from.
  See [HAProxy.md](HAProxy.md)

- max_mime_parts

  Defaults to 1000. There's a potential denial of service in large numbers of
  MIME parts in carefully crafted emails. If this limit is too low for some
  reason you can increase it by setting a value in this file.

- connection_close_message

  Defaults to `closing connection. Have a jolly good day.` can be overrridden with custom text


Changes.md
### Unreleased


#### Changed

- removed dependency on ldap plugins #3399


### [3.0.4] - 2024-08-21

#### Added

- doc: add CONTRIBUTORS #3312
- tls_socket: `config/tls` dir loading is now recursive

#### Changed

- prefix node libs with 'node:' #3359
- .gitignore: add config/me and config/\*.pem
- auth_base: enable disabling constrain_sender at runtime #3298
- auth_base: skip constrain_sender when auth user has no domain #3319
- avg: repackaged as NPM module #3347
- bounce: repackaged plugin as NPM module #3341
- clamd: repackaged plugin as NPM module
- config/plugins: consistent formatting #3359
- connection: check remote is connected before queue #3338
  - improve log message for queue\* hooks, fixes #2998
  - support IPv6 when setting remote.is_private #3295
  - in setTLS, replace forEach with for...of
  - NOTE: remove a handful of 3.0 sunset property names #3315
- contrib/plugin2npm.sh: fix path to package.json #3359
- deps: bump all versions to latest #3303, #3344, #3391
- dkim: repackaged as NPM module #3311
- esets: repackaged as NPM module #3353
- greylist: repackaged as NPM module
- helo.checks: require a successful HELO/EHLO #3352
- new NPM plugin dns-list, repackages dnsbl, dnswl, backscatterer #3313
- when using message-stream, don't send default options #3290
- rcpt_to.host_list: add connection ID to log messages #3322
- line_socket: remove unused callback #3344
- logger: don't load outbound (race condition). Instead, set name property #3322
- logger: extend add_log_methods to Classes (connection, plugins, hmail) #3322
- logger: when logging via `logger` methods, use short names #3322
- logger: check Object.hasOwn to avoid circular deps
- mail_from.resolvable: refactored, leaning on improved net_utils #3322
  - fixes haraka/haraka-net-utils#88
- messagesniffer: repackaged as NPM module
- outbound
  - check for local_mx only when default route is used #3307
  - client_pool: use tls_socket directly (shed line_socket)
  - client_pool: sock.name is now JSON of socket args
  - client_pool.get_client & release_client: arity of 5 -> 2
  - mx_lookup: make it async/await
  - mx_lookup: deleted. Logic moved into net_utils #3322
  - use net_utils.HarakaMx for get_mx parsing #3344
  - emit log message when ignoring local MX #3285
  - pass in config when initiating txn #3315
  - minor es6 updates #3315, #3322
  - logging improvements #3322
    - was: [-] [core] [outbound] Failed to get socket: Outbound connection error: Error: connect ECONNREFUSED 172.16.16.14:25
    - now: [A63B62DF-F3B8-4096-8996-8CE83494A188.1.1] [outbound] Failed to get socket: connect ECONNREFUSED 172.16.16.14:25
  - shorter logger syntax: logger.loginfo -> logger.info
  - remove log prefixes of `[outbound] `, no longer needed
  - delete try_deliver_host. Use net_utils to resolve MX hosts to IPs #3322
  - remove config setting ipv6_enabled #3322
  - remove undocumented use of send_email with arity of 2. #3322
  - encapsulate force_tls logic into get_force_tls #3322
  - es6(async/promise): pre_send_trans_email_respond, process_delivery
- queue/lmtp: refactored for DRY and improved readability #3322
- smtp_client: pass connect_timeout, maybe fixes #3281
- spamassassin: repackaged as NPM module #3348
- style(es6): more for...of loops
- deps: moved attachment, spf, & dkim into optional deps
- doc(Plugins.md): update registry
- doc(Outbound.md): improve GHFM formatting
- remove last vestiges of header_hide_version (long ago renamed)
- server.js: use the local logger methods
  - es6(async): \_graceful, get_smtp_server, setup_smtp_listeners
  - replace async.eachLimit with Promise.all batches
- status: replace async.map with Promise.allSettled
- get Haraka version from utils.getVersion (which includes git id if running from repo)
- tls_socket: remove secureConnection. Fixes #2743
  - getSocketOpts is now async
  - parse_x509 is now async
  - shed dependency on caolin/async & openssl-wrapper
  - get_certs_dir is now async
    - completely refactored.
- transaction: init with conn.init_transaction, always pass in cfg #3315
- test: add a connection.response test case with DSN #3305
- test: convert test runner to mocha
- test: rename tests -> test (where test runner expect) #3340

#### Fixed

- fix(logger): refactor add_log_methods, don't set extra `loglog*` names
- doc(connection): update rfc7001 URL
- fix(bin/haraka): list NPM installed plugin #3310
- fix(bin/haraka): get hook list from doc/Plugins #3306
- fix(outbound): call cb even if no MX is found #3294
- fix(helo.checks): declare reject.literal_mismatch as boolean #3293
- fix(outbound): allow LHLO over insecure socket if TLS is forced #3278
- fix(outbound): include return path param SMTPUTF8 when required #3289
- fix(outbound): replace empty Message-ID header #3288
- fix(outbound): don't send SNI servername when connecting to an IP
- fix(outbound): chown queue dir after creation #3291
- fix(server): async endpoint.bind() and await in server.js #3366
- fix(outbound): get_mx DNS error handling #3376

### [3.0.3] - 2024-02-07

#### Added

- feat(auth_vpopmaild): when outbound, assure the envelope domain matches AUTH domain #3265
- doc(Plugins.md): add pi-queue-kafka #3247
- feat(rabbitmq_amqplib): configurable optional queue arguments #3239
- feat(clamd): add x-haraka-virus header #3207

#### Fixed

- Fix: add empty string as param to .join() on bounce. #3237
- Update links in documentation #3234
- fix(ob/hmail):Add filename to the error for easy debugging
- fix(ob/queue): Ignore 'error.' prefixed files in the queue because corrupted

#### Changed

- docs(outbound): remove example of outbound_ip #3253
- transaction: simplify else condition in add_data #3252
- q/smtp_forward: always register get_mx hook #3204
- dep(pi-es): bump version to 8.0.2 #3206
- dep(redis): bump version to 4.6.7 #3193
- dep(pi-spf): bump version to 1.2.4
- dep(net-utils): bump version to 1.5.3
- dep(pi-redis): bump version to 2.0.6
- dep(tld): bump version to 1.2.0

outbound/config.js
"use strict";

const config = require("haraka-config");
const logger = require("../logger");

exports.name = "outbound/config";

function load_config() {
    const cfg = (exports.cfg = config.get(
        "outbound.ini",
        {
            booleans: [
                "-disabled",
                "-always_split",
                "+enable_tls",
                "-local_mx_ok",
            ],
        },
        () => {
            load_config();
        },
    ).main);

    // legacy config file support. Remove in Haraka 4.0
    if (!cfg.disabled && config.get("outbound.disabled")) {
        cfg.disabled = true;
    }
    if (!cfg.enable_tls && config.get("outbound.enable_tls")) {
        cfg.enable_tls = true;
    }
    if (!cfg.temp_fail_intervals) {
        cfg.temp_fail_intervals = config.get("outbound.temp_fail_intervals");
    }
    if (!cfg.maxTempFailures) {
        cfg.maxTempFailures = config.get("outbound.maxTempFailures") || 13;
    }
    if (!cfg.concurrency_max) {
        cfg.concurrency_max = config.get("outbound.concurrency_max") || 10000;
    }
    if (!cfg.connect_timeout) {
        cfg.connect_timeout = 30;
    }
    if (!cfg.received_header) {
        cfg.received_header =
            config.get("outbound.received_header") || "Haraka outbound";
    }

    exports.set_temp_fail_intervals();
}

exports.set_temp_fail_intervals = function () {
    // Set the outbound temp fail intervals (retry times) using the following rules:
    //   1) temp_fail_intervals takes precedence over maxTempFailures if both are specified
    //   2) if temp_fail_intervals is not specified or is illegally specified, then initialize
    //      it with the equivalent times of maxTempFailures using the original 2^N formula
    //   3) the word "none" can be specified if you do not want to retry a temp failure,
    //      equivalent behavior of specifying maxTempFailures=1
    const { cfg } = this;

    // Fallback function to create an array of the original retry times
    function set_old_defaults() {
        cfg.temp_fail_intervals = [];
        for (let i = 1; i < cfg.maxTempFailures; i++) {
            cfg.temp_fail_intervals.push(2 ** (i + 5));
        }
    }

    // Helpful error function in case of parsing failure
    function error(i, msg) {
        logger.error(
            exports,
            `temp_fail_intervals syntax error parsing element ${i}: ${msg}`,
        );
        logger.warn(
            exports,
            "Setting outbound temp_fail_intervals to old defaults",
        );
        set_old_defaults();
    }

    // If the new value isn't specified, then create the old defaults
    if (!cfg.temp_fail_intervals) {
        return set_old_defaults();
    }

    // If here then turn the text input into an expanded array of intervals (in seconds)
    // i.e, turn "1m,5m*2,1h*3" into [60,300,300,3600,3600,3600]
    // Parse manually to do better syntax checking and provide better failure messages
    const times = [];
    let input = cfg.temp_fail_intervals.replace(/\s+/g, "").toLowerCase();
    if (input.length === 0) return error(0, "nothing specified");
    if (input === "none") {
        cfg.temp_fail_intervals = [];
        return;
    }
    input = input.split(",");

    for (let i = 0; i < input.length; i++) {
        const delay = input[i].split("*");
        if (delay.length === 1) delay.push(1);
        else if (delay.length === 2) delay[1] = Number(delay[1]);
        else return error(i, "too many *");
        if (!Number.isInteger(delay[1]))
            return error(i, "multiplier is not an integer");

        if (delay[0].length < 2) error(i, "invalid time span");
        const symbol = delay[0].charAt(delay[0].length - 1);
        let num = Number(delay[0].slice(0, -1));
        if (isNaN(num)) return error(i, "invalid number or symbol");

        switch (symbol) {
            case "s":
                // do nothing, this is the base unit
                break;
            case "m":
                num *= 60;
                break;
            case "h":
                num *= 3600;
                break;
            case "d":
                num *= 86400;
                break;
            default:
                return error(i, "invalid time span symbol");
        }
        // Sanity check (what should this number be?)
        if (num < 5)
            return error(i, "delay time too small, should be >=5 seconds");
        for (let j = 0; j < delay[1]; j++) {
            times.push(num);
        }
    }

    // One last check, just in case...should never be true
    if (times.length === 0) return error(0, "unexpected parsing result");

    // If here, success, so actually store the calculated array in the config
    cfg.temp_fail_intervals = times;
};

load_config();
